feat: add lab for setting up RAG with KAITO on AKS (#81)
Azure-Samples/aks-labs: docs/ai-workloads-on-aks/kaito-ragengine.mdx

---
title: Build RAG applications with KAITO RAGEngine
---

import Prerequisites from "../../src/components/SharedMarkdown/_prerequisites.mdx";

Retrieval Augmented Generation (RAG) is a powerful technique that combines the strengths of large language models (LLMs) with external knowledge sources. This approach enables more accurate and contextually relevant responses by retrieving information from knowledge bases or databases and using it to augment the LLM's output.

This workshop will guide you through setting up an easy-to-use RAG solution using KAITO on Azure Kubernetes Service (AKS).

> The [KAITO](https://github.com/kaito-project/) project provides Kubernetes operators that aims to simplify the deployment and management of AI tooling on Kubernetes. With the [v0.5.0 release](https://github.com/kaito-project/kaito/releases/tag/v0.5.0), KAITO has introduced a new RAGEngine operator that makes it simple to deploy and manage RAG workloads on AKS.

Reviewer: ```suggestion
The [KAITO](https://github.com/kaito-project/kaito) project provides Kubernetes operators that aims to simplify the deployment and management of AI tooling on Kubernetes. With the [v0.5.0 release](https://github.com/kaito-project/kaito/releases/tag/v0.5.0), KAITO has introduced a new RAGEngine operator that makes it simple to deploy and manage RAG workloads on AKS. Common uses of RAG in AI-enabled applications include:

* Customer support aid that answers FAQs using support manuals or ticket history
* Codebase PR reviewer automating feedback based on files, comments and documentation
* Healthcare doc assistant that ensures secure inference with responses grounded in vetted medical references
```

[KAITO RAGEngine](https://github.com/kaito-project/kaito/blob/main/website/docs/rag.md) is another Kubernetes custom resource that simplifies the deployment and management of Retrieval Augmented Generation (RAG) workloads. It allows you to easily deploy RAG applications and workflows that can retrieve information from external knowledge sources and use it to enhance the responses generated by large language models (LLMs).

In a typical RAG workflow, you have a knowledge base (like a database or an index) and an LLM. Data is often indexed in a vector database, which allows for efficient retrieval of relevant information based on the user's query. This requires a few key components:

- **Knowledge Base**: A source of information that can be indexed and queried to retrieve relevant data.
- **Vector Database**: A database that stores the indexed data in a format that allows for efficient similarity search.
- **Embedding model**: A model that converts text into vector representations, which are then stored in the vector database.
- **Inference Service**: A service that provides access to a large language model (LLM) for generating responses based on the retrieved information.

> With KAITO RAGEngine the deployment of these components are simplified by providing a single CRD that encapsulates the necessary configurations for the vector database, embedding model, and inference service. By default, it uses the [Faiss](https://faiss.ai/) vector database and the [BAAI general embedding (bge)](https://huggingface.co/BAAI/bge-small-en-v1.5) model, but it can be configured to use other embedding models as needed.

Reviewer: ```suggestion
With KAITO RAGEngine, the deployment of these components is simplified by providing a single CRD that encapsulates the necessary configurations for the vector database, embedding model, and inference service. By default, it uses the [Faiss](https://faiss.ai/) vector database and the [BAAI general embedding (bge)](https://huggingface.co/BAAI/bge-small-en-v1.5) model, but it can be configured to use other embedding models as needed.
```

Here is a high-level overview of the RAGEngine architecture:

![KAITO RAGEngine Architecture](../ai-workloads-on-aks/assets/kaito-ragengine/ragarch.png)
Source: [https://github.com/kaito-project/kaito](https://github.com/kaito-project/kaito)

:::info

As you build your RAG application, instead of calling the inference service directly, you will interact with the `/query` endpoint of the RAGEngine instance. This endpoint will handle the retrieval of relevant information from the knowledge base, pass it to the inference service, and return the generated response.

:::

## Deploy RAGEngine

When you applied the Terraform configuration earlier, it also installed the [KAITO RAGEngine Helm Chart](https://github.com/kaito-project/kaito/tree/main/charts/kaito/ragengine) on your AKS cluster. So you can deploy a RAGEngine instance by deploying a single YAML file that defines the RAGEngine configuration.

> Run the following command to deploy a RAGEngine instance named `ragdemo` using the bge model and the phi-4-mini-instruct inference service:

Reviewer: ```suggestion
Run the following command to deploy a RAGEngine instance named `ragdemo` using the bge model and the [phi-4-mini-instruct](https://github.com/kaito-project/kaito/blob/main/examples/inference/kaito_workspace_phi_4_mini.yaml) inference service:
```

:::info

> You can customize the `instanceType` to use a different Azure VM size based on your requirements and available quota. The `Standard_D8s_v4` instance type is a good starting point for development and testing purposes, but you may want to use a larger instance type for production workloads or more demanding applications.

Reviewer: ```suggestion
You can customize the `instanceType` to use a different Azure VM size based on your requirements and available quota. The `Standard_D8s_v4` instance type is a **cost-effective** starting point for development and testing purposes, but you may want to use a larger instance type for production workloads or more demanding applications.
```

Here is an example of what the response might look like:

```json
{
  "response": " Yes, there is a Seashell Snuggle Bed available from Contoso Pet Supply. It's priced at $19.99. It's shaped like a seashell and designed to provide comfort and relaxation for your cat. You can find it [here](http://<STORE_IP>/product/7).\n\n\n",
  "source_nodes": [
  ...
  ]
  ...
}
```

Also included in the response is a `source_nodes` array that contains the documents and relevancy scores that were retrieved from the vector database and used to generate the response. This can be useful for debugging and understanding how the RAGEngine is generating responses based on the indexed data.

## Summary

Congratulations! You've successfully built a complete RAG solution using KAITO on Azure Kubernetes Service. In this workshop, you:

- **Deployed infrastructure** using Terraform to create an AKS cluster with KAITO operators
- **Set up AI workloads** by deploying a phi-4-mini-instruct model workspace and RAGEngine instance
- **Indexed real data** from a sample e-commerce application into a vector database
- **Queried intelligently** to get contextual, source-backed responses from your knowledge base

> With KAITO, the complexity of managing vector databases, embedding models, and inference services is abstracted away, letting you focus on building applications rather than infrastructure.

Reviewer: ```suggestion
With KAITO, the complexity of managing vector databases, embedding models, and inference services is abstracted away, letting you focus on building applications rather than infrastructure. By default, your models, logs, and data store(s) are **secured** with in-cluster RAG, also preventing exposure of real-time sensitive data to external LLM services.
```
